cirq-core/cirq/devices/grid_qubit.py

```diff
@@ -29,9 +29,43 @@
 class _BaseGridQid(ops.Qid):
     """The Base class for `GridQid` and `GridQubit`."""
 
-    def __init__(self, row: int, col: int):
-        self._row = row
-        self._col = col
+    _row: int
+    _col: int
+    _dimension: int
+    _hash: Optional[int] = None
+
+    def __getstate__(self):
+        # Don't save hash when pickling; see #3777.
+        state = self.__dict__
+        if "_hash" in state:
+            state = state.copy()
+            del state["_hash"]
+        return state
+
+    def __hash__(self) -> int:
+        if self._hash is None:
+            self._hash = hash((self._row, self._col, self._dimension))
```

Comment:
We could, but since construction is more common than hashing I think it would be good to keep the constructor as fast as possible. I played around a bit with adding a `__new__` method so we can actually cache instances instead of allocating new ones each time the constructor is called, but that interacts with pickling in strange ways that I haven't worked out yet.

Reply:
Here's a comparison of timings with lazy hashing (this PR) vs eager hashing where as you suggest the hash is computed in the constructor and there's no conditional in `__hash__`. It does indeed make hashing faster, but slows down the constructor quite a bit:

| Operation | Lazy Hash [ns] | Eager Hash [ns] | t_eager / t_lazy |
| ------ | ------ | ------ | ------ |
| `%timeit cirq.GridQubit(1, 2)` | 294 | 430 | 1.46x |
| `%timeit hash(cirq.GridQubit(1, 2))` | 652 | 589 | 0.90x |
| `q = cirq.GridQubit(1, 2); %timeit hash(q)` | 163 | 135 | 0.83x |
| `%timeit cirq.LineQubit(3)` | 254 | 369 | 1.45x |
| `%timeit hash(cirq.LineQubit(3)` | 560 | 506 | 0.90x |
| `q = cirq.LineQubit(3); %timeit hash(q)` | 163 | 144 | 0.88x |
| `%timeit cirq.NamedQubit("abc")` | 249 | 361 | 1.45x |
| `%timeit hash(cirq.NamedQubit("abc")` | 568 | 502 | 0.88x |
| `q = cirq.NamedQubit("abc"); %timeit hash(q)` | 159 | 137 | 0.86x |

So eager hashing speeds up the `__hash__` calls by between 10-20%, but slows down the constructor by almost 50%.

Reply:
Ok, thanks for the thorough analysis.  This makes sense.